### DISCUSSION TITLE
manage storage files from sql

### DISCUSSION THREAD

**Discussion by zakaria-chahboun** [supabase/supabase#3124] [Questions]
I have a form which has some text inputs and image file

* I uploaded the image to the storage by JavaScript API
* I got the storage url
* Then i created a Postgres function to fill the database properly (with image URL as a parameter of course)

My question is: How i can revoke the image from storage in case i have an exception in my Postgres function!

Thanks!

**Comment by zernonia** [comment]
Hi, you can manage your storage files from sql. The `storage` schema has `bucket` and `objects` table for you to query.

You can try the following command and see all the items in the storage:
```sql
select * from storage.objects 
```

Each row has it's own `bucket_id` and `name`, which could help you identify which image URL to remove.
So, in your postgres function, maybe you can do this in case of an exception.
```sql
delete from storage.objects where bucket_id = 'products' and name = 'testing.png'
```

*** 

If you afraid this only remove the entry from the table, and doesn't actually free up the Storage space, you can check `Storage space` at `Home`, before insert image, after insert the image, finally after the query that remove the image. You will see the `Storage space` is the same as before inserting the image.


Hope this help! 😊

**Reply by zakaria-chahboun** [reply]
@zernonia Thanks :+1:

**Reply by Vinzent03** [reply]
Thanks for the tip, but do you know for sure how that `Storage space` is calculated? It's possible it's only counted by the size value in the metadata column.

**Reply by zernonia** [reply]
@Vinzent03 @zakaria-chahboun  I'm so sorry for this `incorrect answer`. As the matter of fact, this method only remove the metadata in the database, and not the actual file. I've confirmed with Supabase team regarding this. 

Thus, the correct way to delete **Storage file** is to call `pg_http`  `DELETE` to your Supabase Storage Endpoint.

**Reply by Vinzent03** [reply]
Ah okay, but even the count under "Billing and usage" is wrong. That's strange and shouldn't be the case, but that's another topic. Do you have a snippet? Never worked with http from postgresql so far. Should I use the `service_role secret` for authentication?

**Reply by GaryAustin1** [reply]
I had just asked this question on the discord channel.   I could have thousands of notes per user each associated with a file and when user leaves/deleted need to get rid of those.  Had planned to just run a trigger function on delete of a note row to delete its attached file(s).  This seems like a major hole.

**Reply by zernonia** [reply]
I'm sorry I've yet to try out the `pg_http` , I was informed by the team that this should be the way.
But, I believe this issue could be resolved when Supabase Function is available.

**Reply by GaryAustin1** [reply]
I'm going to drop this here.  Warning from a beginner on Supabase and postgres, but finally got delete to work using the http extension.  I used two functions for simplicity.  The first is a trigger function on delete than then calls a delete_file function with bucket and path from the "old" variable provided by delete trigger.  I don't like having to hard code api secret code and api url so deep in a function, but don't know how to resolve that.

create or replace function public.delete_file(bucket varchar, pathname varchar) 
returns varchar as $$
declare
   myurl varchar := 'YOUR API URL';
   service_role_key varchar := 'YOUR SERVICE KEY';
  url varchar := myurl||'/storage/v1/object/'||bucket||'/'||pathname;
begin
return (
select status FROM http((
          'DELETE',
            url,
           ARRAY[http_header('authorization','Bearer '||service_role_key)],
           NULL,
           NULL
        )::http_request)
);
end;
$$ language plpgsql security definer;

Good luck.
Edit  I clearly don't know how to format posts in Github.....

**Reply by zernonia** [reply]
@GaryAustin1  Thank you for this code, I haven't tested it but it looks good to me. As I mentioned earlier, let's hope that Supabase Function able to achieve this without us hardcoding the `API KEY` and `SERVICE KEY`. Fingers crossed 🤞🏼

**Reply by zakaria-chahboun** [reply]
@GaryAustin1 Thanks for the code man! :clap:

**Reply by devhandler** [reply]
@GaryAustin1  "The first is a trigger function" could you share the trigger func?

**Reply by GaryAustin1** [reply]
@devhandler This might be overkill depending on what you want to do, but should give you some ideas...
https://github.com/GaryAustin1/supa-file-helper

**Reply by devhandler** [reply]
@GaryAustin1 
the api doc shows it uses remove() to delete files from buckets and it doesn't share any other methods to delete files.
![image](https://github.com/supabase/supabase/assets/133625693/4eca1548-daac-4d7a-ae5c-667c83af4fe9)

in your code, seems you use http delete method. how do you know it would work and where did you get the related api doc?

**Reply by GaryAustin1** [reply]
https://supabase.github.io/storage-api/#/
Is the rest API for storage (you call from http).
.remove() is a javascript call, that if you look at the network request it generates will be an http delete request.
You could also use an edge function to do the operations using supabase-js if desired.

**Reply by ndr3svt** [reply]
I have a question all, 
is this truly deleting the files from the physical storage?

`delete from storage.objects where bucket_id = 'products' and name = 'testing.png'`

I know you kind of mention it, but it is not very explicitly documented in supabase storage ... besides the remove() function in js client , but that leaves us with few options to use other methods

I fin the js client a bit limiting, perhaps somebody has found a work around?

**Reply by GaryAustin1** [reply]
@ndr3svt That will NOT delete the storage file.  It will leave that space in your s3 forever orphaned.  You must use the JS or REST API calls to the storage API.

**Reply by ndr3svt** [reply]
@GaryAustin1  ok, I got it :) thanks for the hint

I'm trying to build a worker that does some garbage collection from assets which got uploaded but their associated entity didn't get saved for some reason, ie crash, user closes the browser, etc
but as a pg_cron task, I don't know how can we invoke the REST API to the storage API, do you know if there is a way to do that? or should I use something else like edge functions?

**Reply by GaryAustin1** [reply]
https://github.com/GaryAustin1/supa-file-helper
In this I use the REST API with both pg_net and http extensions to delete files with cron from deleted user notes that had attached files.
It should give you some ideas.
An edge function would also work.

**Reply by ndr3svt** [reply]
great thank you , I'll take a look at pg_net and the http extensions

**Comment by zakaria-chahboun** [comment]
The supabase storage return the full path like *`(public/products/feca4889.png)`*
I stored this full path in table like this:
|id|product|image|
|----|----|----|
|1|orange|public/products/feca4889.png|

Now. I have to retrieve the `bucket_id` and the `name` of my product image from the full path?
like that:
```SQL
delete from storage.objects where bucket_id = 'public' and name = 'products/feca4889.png'
```
To retrieve the `bucket_id`  from the full path:
```SQL
select split_part('public/products/feca4889.png', '/', 1) as bucket_id;
```
|bucket_id|
|-----------|
|public|

To retrieve `name`  from the full path:
```SQL
select array_to_string((string_to_array('public/products/feca4889.png', '/'))[2:], '/') as name;
```
|name|
|-----------|
|products/feca4889.png|

@zernonia  is this the only solution?

**Reply by EskelCz** [reply]
Thanks, but how would I delete a folder? (with contents, from sql)

**Reply by GaryAustin1** [reply]
You would have to loop thru all the objects in the table with that folder as part of their path name (or using the pathtoken array column) and then do the http call to delete each one.

**Reply by EskelCz** [reply]
Meanwhile I asked GPT-3 and got this
```
CREATE OR REPLACE FUNCTION delete_folder_with_contents(p_folder_path text)
RETURNS VOID AS $$
DECLARE
    file_path text;
BEGIN
    FOR file_path IN (SELECT name FROM storage.objects WHERE name LIKE p_folder_path || '%')
    LOOP
        DELETE FROM storage.objects WHERE name = file_path;
    END LOOP;
END;
$$ LANGUAGE plpgsql;
```
Seems like it's working :) Thanks to you both

**Reply by GaryAustin1** [reply]
@EskelCz  That is very bad!!!!!
That leaves all the files in s3 storage orphaned.  You just deleted the "pointers to them".  Never delete objects directly from schemas you did not create.
You have to use a use http calls to the Storage API to do the deletes, like I showed here:  https://github.com/supabase/supabase/discussions/3124#discussioncomment-1399576

ChatGPT gives very confident answers that many times are wrong, especially with newer stuff like Supabase as it has had no access to the internet in over a year (maybe 2)

**Reply by EskelCz** [reply]
I see. Ok, thanks a lot. :) I got confused because this post was marked as the answer and wasn't disputed. Seemed like an easier code too.
Anyway I thrown your example there as well and got this formatted function. Seems legit.
```
CREATE OR REPLACE FUNCTION public.delete_folder(p_bucket varchar, p_folder_path varchar)
RETURNS varchar AS $$
DECLARE
    myurl varchar := 'YOUR API URL';
    service_role_key varchar := 'YOUR SERVICE KEY';
    file_path varchar;
    url varchar;
    result varchar;
BEGIN
    FOR file_path IN (SELECT name FROM storage.objects WHERE name LIKE p_folder_path || '%')
    LOOP
        url := myurl || '/storage/v1/object/' || p_bucket || '/' || file_path;
        SELECT status
        INTO result
        FROM http(('DELETE', url, ARRAY[http_header('authorization', 'Bearer ' || service_role_key)], NULL, NULL)::http_request);
        RAISE NOTICE 'Deleting file %, status %', file_path, result;
    END LOOP;
    RETURN 'Success';
END;
$$ LANGUAGE plpgsql SECURITY DEFINER;
```

This function takes the p_bucket and p_folder_path as input, and deletes all objects (files) in the folder by calling the Storage API using the http extension. The function loops through all objects in the folder, constructs the URL to delete each object, and deletes it using the DELETE method. The function returns a success message if all objects in the folder are deleted successfully.

Note that you need to replace 'YOUR API URL' and 'YOUR SERVICE KEY' with the actual values for your Supabase setup. Also, the RAISE NOTICE statement is used to log the status of each file deletion, which can be helpful for debugging purposes.

**Reply by GaryAustin1** [reply]
The OP selected their own answer before the others came along.  Not sure how that ever gets corrected...

**Reply by bennik88** [reply]
Any updates on this in 2023? Is there any better approach to deleting files on an auth.user delete trigger than hardcoding the keys like above?
And does anyone know how to delete files that have only been deleted over SQL 
```sql
delete from storage.objects where bucket_id
``` 
function? So files that still exist in the bucket but seem to have no reference in `storage.objects`.

**Reply by GaryAustin1** [reply]
@bennik88 As far as I can tell they are orphaned in s3 storage.  There is no way to delete them once you delete the storage.objects rows.

https://github.com/GaryAustin1/supa-file-helper

This has some automated code but still needs delete trigger.   Just delays the actual storage calls to cron.

**Reply by bennik88** [reply]
Thank you, that is bad news that means the project has to be deleted and a new one created to get rid of the orphaned files. Seems like the abstraction from Supabase is not doing a great job in this case :/ . I even tried to delete the bucket but it doesn't seem to change anything about the used space.

Your `supa-file-helper` seems like a well-made piece of code. But it feels a little bit scary to use a side hack like this especially as I don't fully understand what's going on behind the scenes and it alters columns in the object.storage table.
I probably rather implement some admin UI to delete these files manually in the front end or use another datastore than supabase.
Sad that there is no better solution for a common task like this. Seems like it's still not production ready.

The below link should be deleted by the Supabase Team as it's very misleading. There is a small note about "Not recommended because this only delete the metadata." but it's not clear what it refers to, and a solution is also not given on this page.
https://supabase-sql.vercel.app/delete-storage-item

**Reply by GaryAustin1** [reply]
If you have feed back for Supabase you need to let them know in issues supabase/supabase for docs.  supabase/storage-api for storage.
I'm not sure they will read this discussion.

**Reply by devhandler** [reply]
@GaryAustin1 so in the storage supabase dash UI, it's all about postgres metadata? while i understand direct sql delete only deletes meta data from postgres, i thought in the UI, you could actually delete the file like directly operating the s3 bucket, but seems i'm wrong. the UI might be powered by just postgres and disconnected from the s3 bucket file management system

**Reply by GaryAustin1** [reply]
The dashboard UI is just a React(I think) app.  It uses both storage-api calls and maybe database calls to the storage.objects and bucket tables.  I'm not sure if it just does .list() or direct storage calls.   But all file operations are done thru storage-api for uploading/downloading/deleting files.  The storage-api handles the storage tables and making the calls to a shared (on hosted instances) s3 bucket.

**Reply by devhandler** [reply]
if the UI can directly manipulate the storage (s3 bucket), then accidentally deleting metadata from postgres might be fine. we can still use UI to delete the orphaned files.   but we don't know if that's the case

anyway, thank you a lot for your help! really appreciated it.

**Reply by GaryAustin1** [reply]
@devhandler The storage-api uses the storage.objects tables to know about the files (not s3).  The dashboard UI also uses the storage-api.  If you delete the metadata, there is no way to delete the s3 storage files currently with either the storage-api or the dashboard UI.  The dashboard will not know about the files anymore if the meta data is gone.  .list() uses the storage.objects table to get the file paths.

**Reply by devhandler** [reply]
oh, ic, sry i didn't get you previously. so storage api only deals with postgrses and s3 is hidden behind with no api exposed to us.

**Reply by GaryAustin1** [reply]
Correct.  The storage api relies on postgres storage.objects table for paths and supabase "buckets" (which are just part of the s3 key and not real s3 buckets).  Then it makes s3 calls to handle the actual file operation with that storage key.

**Comment by devhandler** [comment]
> Thank you, that is bad news that means the project has to be deleted and a new one created to get rid of the orphaned files. Seems like the abstraction from Supabase is not doing a great job in this case :/ . I even tried to delete the bucket but it doesn't seem to change anything about the used space.
> 
> Your `supa-file-helper` seems like a well-made piece of code. But it feels a little bit scary to use a side hack like this especially as I don't fully understand what's going on behind the scenes and it alters columns in the object.storage table. I probably rather implement some admin UI to delete these files manually in the front end or use another datastore than supabase. Sad that there is no better solution for a common task like this. Seems like it's still not production ready.
> 
> The below link should be deleted by the Supabase Team as it's very misleading. There is a small note about "Not recommended because this only delete the metadata." but it's not clear what it refers to, and a solution is also not given on this page. https://supabase-sql.vercel.app/delete-storage-item

you should be able to use the remove() function in js client to delete the real files in s3 as well as remove the metadata in postgres. and as @GaryAustin1 mentioned, you could also use empty api, /bucket/bucketid/empty to remove all real files in s3

**Comment by MadeByDouglas** [comment]
according to this post https://github.com/orgs/supabase/discussions/16393 then the above solution will not work, it just orphans the data, or is that not true?

I'm trying to setup a trigger to delete data in a bucket automatically if the row is deleted, but it seems I should just use the client library to manually delete content instead, but would this not be an ideal use-case for triggers?

I'm surprised this isn't more straightforward, but I'm rather new to postgres and s3 etc.

**Reply by GaryAustin1** [reply]
You have to call the storage api to delete the files.   You can do that with a trigger and http extension. But you can not just delete the row in storage.objects.

**Reply by MadeByDouglas** [reply]
ah ok I see, thanks for clarification. So essentially I would use a supabase webhook right? aren't those basically triggers with the http extension?

**Reply by GaryAustin1** [reply]
I’m not sure you can use a webhook unless you call an edge function to then do the delete. You can’t control the format of the http request with a webhook.

**Comment by MadeByDouglas** [comment]
Oh I see
